internal/worker/worker.go

```diff
@@ -27,98 +24,253 @@ type BlockResult struct {
-		rpc:     rpc,
-		storage: storage,
+		rpc: rpc,
 	}
 }
 
-func (w *Worker) Run(blockNumbers []*big.Int) []BlockResult {
-	var wg sync.WaitGroup
+func (w *Worker) Run(blockNumbers []*big.Int) []WorkerResult {
 	blockCount := len(blockNumbers)
-	resultsCh := make(chan BlockResult, blockCount)
-	for _, blockNumber := range blockNumbers {
-		wg.Add(1)
+	chunks := divideIntoChunks(blockNumbers, w.rpc.BlocksPerRequest.Blocks)
+
+	var wg sync.WaitGroup
+	resultsCh := make(chan []WorkerResult, len(chunks))
 
-		go func(bn *big.Int) {
+	log.Debug().Msgf("Processing %d chunks of blocks", len(chunks))
+	for _, chunk := range chunks {
+		wg.Add(1)
+		go func(chunk []*big.Int) {
 			defer wg.Done()
-			result := w.processBlock(bn)
-			resultsCh <- result
-		}(blockNumber)
+			resultsCh <- w.processBatch(chunk)
+		}(chunk)
 	}
 	go func() {
 		wg.Wait()
 		close(resultsCh)
 	}()
 
-	results := make([]BlockResult, 0, blockCount)
-	for result := range resultsCh {
-		results = append(results, result)
+	results := make([]WorkerResult, 0, blockCount)
+	for batchResults := range resultsCh {
+		results = append(results, batchResults...)
 	}
 
 	return results
 }
 
-func (w *Worker) processBlock(blockNumber *big.Int) BlockResult {
-	log.Debug().Msgf("Processing block %d", blockNumber)
-	block, err := w.fetchBlock(blockNumber)
-	if err != nil {
-		return BlockResult{BlockNumber: blockNumber, Error: fmt.Errorf("error fetching block %d: %v", blockNumber, err)}
+func divideIntoChunks(blockNumbers []*big.Int, batchSize int) [][]*big.Int {
+	if batchSize >= len(blockNumbers) {
+		return [][]*big.Int{blockNumbers}
 	}
+	var chunks [][]*big.Int
+	for i := 0; i < len(blockNumbers); i += batchSize {
+		end := i + batchSize
+		if end > len(blockNumbers) {
+			end = len(blockNumbers)
+		}
+		chunks = append(chunks, blockNumbers[i:end])
+	}
+	return chunks
+}
 
-	logs, err := w.fetchLogs(blockNumber)
+func (w *Worker) processBatch(blockNumbers []*big.Int) []WorkerResult {
+	blocks := w.fetchBlocks(blockNumbers)
+	logs := w.fetchLogs(blockNumbers)
+
+	var traces []TracesFetchResult
+	if w.rpc.SupportsTraceBlock {
+		traces = w.fetchTraces(blockNumbers)
+	}
+
+	return SerializeWorkerResults(blocks, logs, traces)
+}
+
+func (w *Worker) fetchBlocks(blockNumbers []*big.Int) []BlockFetchResult {
+	batch := make([]rpc.BatchElem, len(blockNumbers))
+	blockResults := make([]BlockFetchResult, len(blockNumbers))
+
+	for i, blockNum := range blockNumbers {
+		blockResults[i] = BlockFetchResult{
+			BlockNumber: blockNum,
+		}
+		batch[i] = rpc.BatchElem{
+			Method: "eth_getBlockByNumber",
+			Args:   []interface{}{hexutil.EncodeBig(blockNum), true},
+			Result: new(map[string]interface{}),
+		}
+	}
+
+	err := w.rpc.RPCClient.BatchCallContext(context.Background(), batch)
 	if err != nil {
-		return BlockResult{BlockNumber: blockNumber, Error: fmt.Errorf("error fetching logs for block %d: %v", blockNumber, err)}
+		for i := range blockResults {
+			blockResults[i].Error = err
+		}
```

Comment:
This means if one batch element fails we fail all elements?

Reply:
No, this means if the batch request fails all together, we set the error for every block

Reply:
The doc in the code is telling it differently. The function already sets the error object for the blocks that fails. the err object means one or more blocks failed. https://github.com/ethereum/go-ethereum/blob/80b529ea713d71a27282ae76c2741c8bc5486502/rpc/client.go#L395

Reply:
That is how works. We send a batch request
* request works, all responses are successful -> we set the result for every block
* request works, but not all responses are successful -> we set results for successful blocks and Error for unsuccessful ones (the error is on the element of  `blockResults`)
* (this current case) request does not work and returns a single error -> we propagate that error for every block number